src/ssplit/typedefs.h

```diff
@@ -1,20 +1,46 @@
 #pragma once
+#include <map>
 #include <string>
-#if USE_ABSEIL
-#  include "absl/strings/string_view.h"
-#  include "absl/container/flat_hash_map.h"
+
+// We need the extra level of macro redirection below to make the
+// preprocessor to expand the macro CONVERT_TO_STRING as intended.
+// Uncomment the code block below if you need to debug C++ version
+// settings.
+//
+// #define CONVERT_TO_STRING_(x) #x
+// #define CONVERT_TO_STRING(x) CONVERT_TO_STRING_(x)
+// #pragma message(CONVERT_TO_STRING(__cplusplus))
+
+
+#if defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 6
+// On gcc version 5.x , string_view is provided as
+// std::experimental::string_view
+// Newer versions of clang apparently export GNUC v 4.2, so we need
+// the extra check to exclude the __clang__ scenario.
+#include <experimental/string_view>
+#elif defined(__clang__) && __cplusplus < 201703L
+// Older versions of cmake (e.g. 3.8 on Ubuntu 16.04) don't support
+// CMAKE_CXX_STANDARD 17. On Ubuntu 16.04, clang version also
+// finds string_view here:
+#  include <experimental/string_view>
 #else
-#  include <map>
+#  include <string_view>
 #endif
 
-namespace ug{
-namespace ssplit{
-
-#if USE_ABSEIL
-  typedef absl::string_view string_view;
-  typedef absl::flat_hash_map<std::string, int> prefix_map_t;
+namespace ug {
+namespace ssplit {
+#if defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 6
+typedef std::experimental::string_view string_view;
+#elif defined(__clang__) && __cplusplus < 201703L
+typedef std::experimental::string_view string_view;
 #else
-  typedef std::string_view string_view;
-  typedef std::map<std::string, int, std::less<> > prefix_map_t;
+typedef std::string_view string_view;
```

Comment:
`USE_ABSEIL` is just `USE_ABSEIL_STRING_VIEW` in essence right? Can we have some consistency with [sentencepiece](https://github.com/google/sentencepiece/blob/351600c2971401f4e849147579aa1b5d42f614e1/src/sentencepiece_processor.h#L24-L47)?

I want to avoid conversions and have a clean code eventually in bergamot-translator.

Reply:
> `USE_ABSEIL` is just `USE_ABSEIL_STRING_VIEW` in essence right? Can we have some consistency with [sentencepiece](https://github.com/google/sentencepiece/blob/351600c2971401f4e849147579aa1b5d42f614e1/src/sentencepiece_processor.h#L24-L47)?
> 
> I want to avoid conversions and have a clean code eventually in bergamot-translator.

First, as far as I can tell, bergamot-translator is currently already doing all the necessary conversions. No additional action required.

Second, when push comes to shove, conversions between different string_view implementations are cheap; so cheap in fact, that it is recommended to pass read-only string_views by value rather than by reference (http://www.nuonsoft.com/blog/2018/06/06/c17-stdstring_view/).

Third, why should sentencepiece drive the development of bergamot-translator? Bergamot-translator is currently already using std::string_view. I originally suggested using Sentencepiece as the primary vocab implementation and got booed off the stage immediately. Instead, we are now creating a SentencepieceText instance (including all memory allocation involved), then (unnecessarily, unless being independent of sentencepiece is crucially important to you) copy all kinds of byte offsets into new vectors (memory allocation included), just in order to be "independent of sentencepiece". Can you elaborate why the dirt-cheap conversion between string_view implementations by different 3rd-party components (already implemented) now suddenly is a problem?

Reply:
umm... I just proposed USE_ABSEIL is just USE_ABSEIL_STRING_VIEW, rest is just wishful thinking. Resolving.